src/Features/LanguageServer/Protocol/Handler/Definitions/AbstractGoToDefinitionHandler.cs

```diff
@@ -70,6 +70,12 @@ public AbstractGoToDefinitionHandler(IMetadataAsSourceFileService metadataAsSour
             {
                 // No definition found - see if we can get metadata as source but that's only applicable for C#\VB.
                 var symbol = await SymbolFinder.FindSymbolAtPositionAsync(document, position, cancellationToken).ConfigureAwait(false);
+
+                if (symbol.IsTupleType())
```

Comment:
doc.

not sure hwo i feel about this.  I think i'd prefer if _metadataAsSourceFileService.GetGeneratedFileAsync jsut worked on a tuple type.  so less special casing here.

Reply:
`_metadataAsSourceFileService.IsNavigableMetadataSymbol(symbol)` will not let me reach that code.

```csharp
if (!symbol.Locations.Any(static l => l.IsInMetadata)) // locations are preserved in OriginalDefinition
{
    return false;
}
```

I was thinking of adding this special case to this method but it is used in several places so was not sure if is safe to do so.

1. [src\EditorFeatures\Core\CodeDefinitionWindow\DefinitionContextTracker.cs](https://github.com/dotnet/roslyn/blob/e7cae311ab49148335e969b94fb47f0dffaa95df/src/EditorFeatures/Core/CodeDefinitionWindow/DefinitionContextTracker.cs#L198-L204)
2. [src\EditorFeatures\Core.Wpf\Peek\PeekableItemFactory.cs](https://github.com/dotnet/roslyn/blob/e7cae311ab49148335e969b94fb47f0dffaa95df/src/EditorFeatures/Core.Wpf/Peek/PeekableItemFactory.cs#L82-L95)
3. [src\Features\LanguageServer\Protocol\Handler\Definitions\AbstractGoToDefinitionHandler.cs - Current](https://github.com/dotnet/roslyn/blob/e7cae311ab49148335e969b94fb47f0dffaa95df/src/Features/LanguageServer/Protocol/Handler/Definitions/AbstractGoToDefinitionHandler.cs#L69-L88)
4. [src\VisualStudio\Core\Def\Workspace\VisualStudioSymbolNavigationService.cs](https://github.com/dotnet/roslyn/blob/e7cae311ab49148335e969b94fb47f0dffaa95df/src/VisualStudio/Core/Def/Workspace/VisualStudioSymbolNavigationService.cs#L85-L90)
5. [src\VisualStudio\Xaml\Impl\Implementation\LanguageServer\Handler\Definitions\GoToDefinitionHandler.cs](https://github.com/dotnet/roslyn/blob/e7cae311ab49148335e969b94fb47f0dffaa95df/src/VisualStudio/Xaml/Impl/Implementation/LanguageServer/Handler/Definitions/GoToDefinitionHandler.cs#L173-L189)

Or maybe changing indeed this method and adjusting the references of this method to capture the correct locations and construct the metadata definition item?

Reply:
All of those are ones that i'd want to do the right thing wrt to tuples and ValueTuple, so that makes me want thsi mroe strongly.

Reply:
Note: IsNavigableMetadataSymbol should absolutely be getting to the original symbol (including with tuples, seeing through to the underlying ValueTuple).  That seems like a great place to put things :)

Reply:
I was thinking maybe replacing the symbol with the `OriginalDefinition` for tuple types but there are references of this method that are already doing this, so I'm afraid of duplicate access to original definition.

Maybe something like this:
```diff
// Raw code, it may be simplified/refactored for readability
- if (!symbol.Locations.Any(static l => l.IsInMetadata))
+ if (!(symbol.Locations.Any(static l => l.IsInMetadata) || (symbol.IsValueTuple() && symbol.OriginalDefinition.Locations.Any(static l => l.IsInMetadata))))
{
    return false;
}
```

... and adjusting all those 5 references to ensure the location is correctly retrieved?

Reply:
i would jsut do:

`symbol = symobl.OriginalDefinition;` above the if check.

Reply:
>  so I'm afraid of duplicate access to original definition.

it's not a problem.

Reply:
Ok, then I will it a try. Thanks.

Reply:
you're welcome :)